src/doc/en/developer/coding_basics.rst

```diff
@@ -345,9 +345,25 @@ information. You can use the existing functions of Sage as templates.
 
    The INPUT block describes all arguments that the function accepts.
 
-   1. The type names should be descriptive, but do not have to represent the
-      exact Sage/Python types. For example, use "integer" for anything that
-      behaves like an integer, rather than "int" or "Integer".
+  1. Put type information first and foremost in the function signature
+     (Python / Cython annotations). Treat the signature as the single
+     authoritative place for ordinary type declarations. The INPUT
+     block must not restate obvious types already conveyed there.
+     Instead, use the INPUT block to document the *semantic constraints*
+     and any *non‑obvious* accepted forms:
+
+     - State mathematical or structural conditions: e.g. "prime integer",
+       "nonnegative", "an iterable of vertices", "a square matrix", etc.
+     - Mention coercion behavior, mutability expectations, or protocol
+       requirements ("must implement ``.degree()``", "anything with a
+       ``.parent()``").
+     - Only include explicit type names if they add information not present
+       in the annotation (e.g. symbolic vs numeric, sparse vs dense).
+
+     Avoid redundant phrases like "``n`` -- integer" when the signature reads
+     ``def f(n: int, ...)``; write instead "``n`` -- nonnegative" or
```

Comment:
```suggestion
     ``def f(n: Integer | int, ...)``; write instead "``n`` -- nonnegative" or
```

Should we have this as the recommendation for now and leave adding a type alias for `Integer | int` for follow-up? Or would you prefer to add the type alias in this PR?